### PR TITLE
Update vmbackup to support Azure Managed Identity

### PR DESCRIPTION
**Pull request by justinrush**
### Describe Your Changes

These changes support using Azure Managed Identity for the `vmbackup` utility. It adds two new environment variables:

* `AZURE_USE_DEFAULT_CREDENTIAL`: Instructs the `vmbackup` utility to build a connection using the [Azure Default Credential](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity@v1.5.2#NewDefaultAzureCredential) mode. This causes the Azure SDK to check for a variety of environment variables to try and make a connection. By default, it tries to use managed identity if that is set up.

This will close https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5984

### Checklist

The following checks are **mandatory**:

- [x] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### Testing

However you normally test the `vmbackup` utility using Azure Blob should continue to work without any changes. The set up for that is environment specific and not listed out here.

Once regression testing has been done you can set up [Azure Managed Identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/overview) so your resource (AKS, VM, etc), can use that credential method. Once it is set up, update your environment variables according to the updated documentation.

I added unit tests to the `FS.Init` function, then made my changes, then updated the unit tests to capture the new branches.

I tested this in our environment, but with SAS token auth and managed identity and it works as expected.

### CODE REVIEW

File: lib/backup/azremote/azblob.go [hunk 1]
```diff
@@ -58,32 +65,64 @@ func (fs *FS) Init() error {
...
+		sc, err = service.NewClientFromConnectionString(connString, nil)
 		if err != nil {
 			return fmt.Errorf("failed to create AZBlob service client from connection string: %w", err)
 		}
-	}
-
-	accountName, ok1 := envtemplate.LookupEnv(envStorageAcctName)
-	accountKey, ok2 := envtemplate.LookupEnv(envStorageAccKey)
-	if ok1 && ok2 {
+	case hasAccountName && hasAccountKey:
```
Comment (zekker6) [inline lib/backup/azremote/azblob.go:95]: Show this handle a case when only `envStorageAcctName` is set?
It is possible that user supplied only account name and it seems like this case will not return an error.

File: lib/backup/azremote/azblob.go [hunk 2]
```diff
@@ -383,3 +420,28 @@ func (fs *FS) ReadFile(filePath string) ([]byte, error) {
 
 	return b, nil
 }
+
+// envLookuper is an interface for looking up environment variables. It is
+// needed to allow unit tests to provide alternate values since the envtemplate
+// package uses a singleton to read all environment variables into memory at
+// init time.
+type envLookuper interface {
```
Comment (zekker6) [inline lib/backup/azremote/azblob.go:428]: I would suggest to simplify `envLookuper` to just a func:
```
type envLookuper func(name string) (string, bool)
```
This still allows to perform proper testing, but simplifies code a bit by removing an interface abstraction since interface contains just one method anyway.

File: lib/backup/azremote/azblob_test.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+		"multiple dir prefix / is removed": {
+			Dir:         "//foo/",
+			ExpectedDir: "foo/",
+			ExpectedErr: "failed to detect any credentials",
+		},
+		"suffix is added": {
+			Dir:         "foo",
+			ExpectedDir: "foo/",
+			ExpectedErr: "failed to detect any credentials",
+		},
```
Comment (zekker6) [inline lib/backup/azremote/azblob_test.go:34]: Could you split tests for `Dir` verification and `Client` init into separate functions?
That allows to narrow the test cases scope and makes code easier to read.

File: lib/backup/azremote/azblob_test.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+			if fs.Dir != test.ExpectedDir {
+				t.Errorf("expected dir %q, got %q", test.ExpectedDir, fs.Dir)
+			}
+
+			tlog.MustContain(t, test.ExpectedLogs...)
+		})
+	}
+}
+
+func checkErr(t *testing.T, err error, shouldContain string) {
```
Comment (zekker6) [inline lib/backup/azremote/azblob_test.go:146]: I would suggest to refactor this to avoid the following:
- checking the error mesasge text - replace this with specific errors by type. This will allow to avoid necessity to update tests if the error message will be changed. Seems like errors should be created for the following cases: no credentials, invalid credentials, multiple credentials.
- checking the errors which is returned by underlying Azure library - these tests can start failing after the library updates, so it will require to update tests as well.

**Comment by zekker6** [comment]
Hi @justinrush, thank you for the PR!
Please, see a few comments.

File: lib/backup/azremote/azblob.go (hunk 1 above)
Comment (justinrush) [reply lib/backup/azremote/azblob.go:91]: it will return an error - i added a case to show that here: 92ab76fe5

File: lib/backup/azremote/azblob.go (hunk 2 above)
Comment (justinrush) [reply lib/backup/azremote/azblob.go]: 5085381c0

File: lib/backup/azremote/azblob_test.go (hunk 3 above)
Comment (justinrush) [reply lib/backup/azremote/azblob_test.go]: 93ed9423d

File: lib/backup/azremote/azblob_test.go (hunk 4 above)
Comment (justinrush) [reply lib/backup/azremote/azblob_test.go]: e8cb18dfd

**Comment by justinrush** [comment]
> Hi @justinrush, thank you for the PR! Please, see a few comments.

addressed - LMK if you want me to squash those commits down. I am not sure if ya'll squash commit into `master` or if you pull them all forward from this branch.

**Review by zekker6 — APPROVED** [review]
LGTM
@justinrush see a small follow-up for error handling here - https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6518/commits/862969a2d46c65210e2c8b7e6dbd910276765e6a

File: lib/backup/azremote/azblob.go [hunk 5]
```diff
@@ -41,49 +55,74 @@ type FS struct {
...
-		sc, err = service.NewClientFromConnectionString(cs, nil)
+
+	// not used if connection string is set
+	serviceURL := fmt.Sprintf("https://%s.%s/", accountName, domain)
+
+	switch {
+	// can't specify any combination of more than one credential
+	case moreThanOne(hasConnString, (hasAccountName && hasAccountKey), (useDefault == "true" && hasAccountName)):
+		logger.Errorf("only one of connection string, account name and key, or default credential can be specified")
+		return errNoCredentials
```
Comment (hagen1778) [inline lib/backup/azremote/azblob.go:95]: returned error seems misleading - it will complain about no creds set. In the same time the another error log will be printed with correct message. I don't understand why we can't just return a correct error instead of logging it here?

File: lib/backup/azremote/azblob.go [hunk 6]
```diff
@@ -41,49 +55,74 @@ type FS struct {
...
+	case hasConnString:
+		logger.Infof("Creating AZBlob service client from connection string")
+		sc, err = service.NewClientFromConnectionString(connString, nil)
 		if err != nil {
-			return fmt.Errorf("failed to create AZBlob service client from connection string: %w", err)
+			return fmt.Errorf("failed to create AZBlob service client: %w", err)
```
Comment (hagen1778) [inline lib/backup/azremote/azblob.go:100]: can we put context about used cred types in the error message? If we do, we can remove the Info line above. This is related to the cases below as well. Or am I missing something?

Comment (justinrush) [reply lib/backup/azremote/azblob.go:107]: As a consumer of the functionality I like seeing what path it takes to generate a connection to azure - should it be Debug level?

Not opposed to wrapping the underlying error with context if you think we should do that as well.

Comment (hagen1778) [reply lib/backup/azremote/azblob.go:107]: Understood. Thanks

File: docs/CHANGELOG.md
Comment (hagen1778) [inline docs/CHANGELOG.md:1]: Would you mind reverting unrelated changes?

Comment (justinrush) [reply docs/CHANGELOG.md:1]: just the move of azidentity from indirect to direct (that's just a result of `go mod tidy`)?

if not that, what changes?

Comment (justinrush) [reply docs/CHANGELOG.md:1]: gross, sorry, github was hiding that diff in the files changed activity. Fixed with: 460febcc8f6aa98bfee2b12af0c7cb23deab5bdc
